controller/volume_controller.go

```diff
@@ -3734,50 +3734,49 @@ func (c *VolumeController) processMigration(v *longhorn.Volume, es map[string]*l
 		err = errors.Wrapf(err, "failed to process migration for %v", v.Name)
 	}()
 
-	if !isMigratableVolume(v) {
+	if !util.IsMigratableVolume(v) {
 		return nil
 	}
 
 	log := getLoggerForVolume(c.logger, v).WithField("migrationNodeID", v.Spec.MigrationNodeID)
 
-	// only process if volume is attached and running
-	if v.Spec.NodeID == "" || v.Status.CurrentNodeID == "" || len(es) == 0 {
-		return nil
-	}
-	if v.Status.Robustness != longhorn.VolumeRobustnessDegraded && v.Status.Robustness != longhorn.VolumeRobustnessHealthy {
-		log.Warnf("Skip the migration processing since the volume current robustness is %v", v.Status.Robustness)
-		return nil
-	}
-
 	// cannot process migrate when upgrading
-	if c.isVolumeUpgrading(v) {
+	if isVolumeUpgrading(v) {
 		log.Warn("Skip the migration processing since the volume is being upgraded")
 		return nil
 	}
 
-	// the only time there should be more then 1 engines is when we are migrating or upgrading
-	// if there are more then 1 and we no longer have a migration id set we can cleanup the extra engine
 	if v.Spec.MigrationNodeID == "" {
+		// The volume attachment controller has stopped the migration (if one was ever started). We must clean up any
+		// extra engines/replicas and leave the volume in a "good" state.
+
+		// The only time there should be more then one engines is when we are migrating or upgrading. If there are more
+		// then one and we no longer have a MigrationNodeID set we can cleanup the extra engine.
 		if len(es) < 2 && v.Status.CurrentMigrationNodeID == "" {
-			return nil
+			return nil // There is nothing to do.
+		}
+
+		// The volume is no longer attached or should no longer be attached. We will clean up the migration below by
+		// removing the extra engine and replicas. Warn the user.
+		if v.Spec.NodeID == "" || v.Status.CurrentNodeID == "" {
+			msg := ("Volume migration failed unexpectedly; detach volume from extra node to resume")
+			c.eventRecorder.Event(v, corev1.EventTypeWarning, constant.EventReasonMigrationFailed, msg)
```

Comment:
Conceptual question: is there a rate limit for event emitting?

Reply:
Good question. I looked into this a bit.

- The volume attachment controller creates an EventBroadcaster with the default options.
  https://github.com/longhorn/longhorn-manager/blob/e5701524c9610cb780ef06960d8ede62f25fa7ec/controller/volume_attachment_controller.go#L55
- This leads to the use of an EventCorrelator with default options.
  https://github.com/longhorn/longhorn-manager/blob/e5701524c9610cb780ef06960d8ede62f25fa7ec/vendor/k8s.io/client-go/tools/record/events_cache.go#L477-L507
- The default options allow a source to send a burst of 25 events, but the "bucket" used only refills at a rate of 1 event per 300 seconds.
  https://github.com/longhorn/longhorn-manager/blob/e5701524c9610cb780ef06960d8ede62f25fa7ec/vendor/k8s.io/client-go/tools/record/events_cache.go#L44-L49
- Some more details about how this works.
  https://github.com/longhorn/longhorn-manager/blob/e5701524c9610cb780ef06960d8ede62f25fa7ec/vendor/k8s.io/client-go/tools/record/events_cache.go#L424-L438

TBH, I'm not 100% sure of the effects in relation to your question. It looks to me like the EventBroadcaster will eventually stop broadcasting events from this source (volume controller) for this object (volume) if we try to spam them. And identical events will not appear twice, but instead have their count updated.

In practice, I think the only way we can really hit the rate limit is if we repeatedly fail to pick the new current engine or repeatedly fail to delete the old engine.

Reply:
I inserted a test error just after emitting the event to see what happened. (So, after purposely causing a MigrationFailed event, Longhorn could never recover.)

- Only one event was ever created, with its `count` increasing over time.
- On the Longhorn side, we try (and fail) to `processMigration` four times every 30 seconds. (Every volume queues every thirty seconds and there are a maximum of three retries.)
- On the event side, after a short while, we only update the event once every five minutes. When we do, though, we increase the count by the number of times we called `Event()` in the interim.

```
# Before an interval "rollover"

eweber@laptop:~/longhorn> kl get event pvc-82ebabc8-777d-4939-9198-af33320a379e.17de7c648a6d2cd6
LAST SEEN   TYPE      REASON            OBJECT                                            MESSAGE
4m55s       Warning   MigrationFailed   volume/pvc-82ebabc8-777d-4939-9198-af33320a379e   Volume migration failed unexpectedly; detach volume from extra node to resume

eweber@laptop:~/longhorn> kl get event pvc-82ebabc8-777d-4939-9198-af33320a379e.17de7c648a6d2cd6 -oyaml
apiVersion: v1
count: 245
eventTime: null
firstTimestamp: "2024-07-02T19:34:38Z"
involvedObject:
  apiVersion: longhorn.io/v1beta2
  kind: Volume
  name: pvc-82ebabc8-777d-4939-9198-af33320a379e
  namespace: longhorn-system
  resourceVersion: "120809419"
  uid: 57f11de5-f73c-42a8-88bd-3e72efe0a4a1
kind: Event
lastTimestamp: "2024-07-02T20:05:06Z"
message: Volume migration failed unexpectedly; detach volume from extra node to resume
metadata:
  creationTimestamp: "2024-07-02T19:34:38Z"
  name: pvc-82ebabc8-777d-4939-9198-af33320a379e.17de7c648a6d2cd6
  namespace: longhorn-system
  resourceVersion: "120818848"
  uid: d3fc0f66-b4c4-47c6-835a-ef1fbb9f1933
reason: MigrationFailed
reportingComponent: longhorn-volume-controller
reportingInstance: ""
source:
  component: longhorn-volume-controller
type: Warning

# After an interval "rollover"

eweber@laptop:~/longhorn> kl get event pvc-82ebabc8-777d-4939-9198-af33320a379e.17de7c648a6d2cd6
LAST SEEN   TYPE      REASON            OBJECT                                            MESSAGE
1s          Warning   MigrationFailed   volume/pvc-82ebabc8-777d-4939-9198-af33320a379e   Volume migration failed unexpectedly; detach volume from extra node to resume

eweber@laptop:~/longhorn> kl get event pvc-82ebabc8-777d-4939-9198-af33320a379e.17de7c648a6d2cd6 -oyaml
apiVersion: v1
count: 285
eventTime: null
firstTimestamp: "2024-07-02T19:34:38Z"
involvedObject:
  apiVersion: longhorn.io/v1beta2
  kind: Volume
  name: pvc-82ebabc8-777d-4939-9198-af33320a379e
  namespace: longhorn-system
  resourceVersion: "120809419"
  uid: 57f11de5-f73c-42a8-88bd-3e72efe0a4a1
kind: Event
lastTimestamp: "2024-07-02T20:10:06Z"
message: Volume migration failed unexpectedly; detach volume from extra node to resume
metadata:
  creationTimestamp: "2024-07-02T19:34:38Z"
  name: pvc-82ebabc8-777d-4939-9198-af33320a379e.17de7c648a6d2cd6
  namespace: longhorn-system
  resourceVersion: "120820386"
  uid: d3fc0f66-b4c4-47c6-835a-ef1fbb9f1933
reason: MigrationFailed
reportingComponent: longhorn-volume-controller
reportingInstance: ""
source:
  component: longhorn-volume-controller
type: Warning
```

I am not inclined to make changes here because:

- We should not be able to remain in this state indefinitely.
- Even if we do, the above client-side rate limiting will keep us from hammering the API server (eventually).
- It seems there are other locations in the volume controller at which we would also generate events similarly if some unknown bug surfaced.
- The solution (I think) would have to be either an additional in-memory backoff in addition to the mechanism built into the event client.